src/awkward1/operations/structure.py

```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
             + awkward1._util.exception_suffix(__file__)
         )
 
-    batch = [contents[0]]
-    for x in contents[1:]:
-        if batch[-1].mergeable(x, mergebool=mergebool):
-            batch.append(x)
-        else:
-            collapsed = batch[0].mergemany(batch[1:])
-            batch = [collapsed.merge_as_union(x)]
+    if axis < 0:
+        raise ValueError(
+            "the 'axis' of concatenate must be non-negative"
+            + awkward1._util.exception_suffix(__file__)
+        )
+    elif axis == 0:
+        batch = [contents[0]]
+        for x in contents[1:]:
+            if batch[-1].mergeable(x, mergebool=mergebool):
+                batch.append(x)
+            else:
+                collapsed = batch[0].mergemany(batch[1:])
+                batch = [collapsed.merge_as_union(x, 0)]
+
+        out = batch[0].mergemany(batch[1:])
+    else:
+        length = len(contents[0])
+        if any(len(lst) != length for lst in contents[1:]):
+            raise ValueError("all arrays must have the same length for concatenate in axis > 0"
+                + awkward1._util.exception_suffix(__file__)
+            )
+
+        def getfunction(inputs, depth):
+            if depth == axis:
+                arr = inputs
+                batch = [arr[0]]
+                for x in arr[1:]:
+                    collapsed = batch[0].mergemany_as_union(batch[1:], 1)
+                    batch = [collapsed.merge_as_union(x, 1)]
+                out = batch[0].mergemany_as_union(batch[1:], 1)
```

Comment:
I don't understand this: why is `mergemany_as_union` and `merge_as_union` being called once per array? Why shouldn't this implementation just be the following?

```python
out = inputs[0].mergemany_as_union(inputs[1:], 1)
```

Reply:
done

Reply:
I lost a huge comment here...

Looking at the original, I can see what it was doing:

```python
    batch = [contents[0]]
    for x in contents[1:]:
        if batch[-1].mergeable(x, mergebool=mergebool):
            batch.append(x)
        else:
            collapsed = batch[0].mergemany(batch[1:])
            batch = [collapsed.merge_as_union(x)]

    out = batch[0].mergemany(batch[1:])
```

It was accumulating batches of mutually `mergeable` arrays into a Python list (which does not scale with the size of the data) and then either calling `mergemany` on them all once at the end or once as soon as a non-mergeable array was encountered. Thus, the algorithm was _O(n)_ in the total length of the array: nothing was double-merged.

I haven't checked your implementation to see if it has the same characteristics. We'd want it to be able to batch mutually `mergeable` arrays and only `mergemany` them once (but at a deeper `axis`) and we'd want it to not attempt to `mergemany` on non-mergeable arrays. Keep in mind that the `mergemany` method raises RuntimeErrors (i.e. our internal errors) if you try to merge anything that can't be merged. If we don't compare explicitly verify that two arrays are `mergeable` before trying to merge them, it's a bug.

In the tests, can you verify that all combinations of the following arrays containing numbers and records (at `axis=1`) concatenate correctly?

```python
numbers = [
    awkward1.Array([[1.1, 2.2, 3.3], [], [4.4, 5.5]]),
    awkward1.Array([[10, 20], [30], [], [40, 50, 60]]),
    awkward1.Array(numpy.array([[101, 102, 103], [201, 202, 203]])),
]
records = [
    awkward1.Array([[{"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}], [], [{"x": 2.2, "y": [1, 2]}]]),
    awkward1.Array([[{"x": 3.3, "y": [1, 2, 3]}], [{"x": 4.4, "y": [2, 1]}, {"x": 5.5, "y": [1]}]]),
]
```

That is, lists like `numbers + records`, `numbers[:1] + records + numbers[1:]`, etc.? The two constraints that must be upheld are

   1. It must call `mergemany` exactly once on a contiguous run of same-type arrays (all numbers or all records).
   2. It must not call `mergemany` on different-type arrays.

You've also implemented a `mergemany_as_union`, which would be nice to call only once, rather than for every pair. Note that a UnionArray is `mergeable` with anything, since the merge would only make a UnionArray with more unique tags.

Reply:
done